### ISSUE TITLE
systemd-udevd should close fd when running daemonized outside of systemd

### ISSUE THREAD

**Issue by Whissi** [systemd/systemd#190]
Bug-Gentoo: https://bugs.gentoo.org/show_bug.cgi?id=547916

Since commit 5c67cf2774a8b964f4d7cd92a4c447da81ac6087 _systemd-udevd_ doesn't close std{in,out,err} anymore, even when running _systemd-udevd_ daemonized (`systemd-udevd --daemon`).

As result you cannot use Python's subprocess module to start/restart _systemd-udevd_ like you could do in <=systemd-216 for example. That's how I run into that problem:

I tried salt (a configuration management tool) to restart _systemd-udevd_, see https://github.com/saltstack/salt/issues/14957#issuecomment-68957648

According to http://www.freedesktop.org/software/systemd/man/daemon.html (9) a daemonized service should connect _/dev/null_ to standard input, output, and error. Can we do the same with _systemd-udevd_ when running daemonized?

PS: You maybe wondering why I am posting this issue here if the Gentoo bug is already closed and marked as fixed:
We thought that running _systemd-udevd_ in foreground would be a workaround but we now run into another problem (https://bugs.gentoo.org/show_bug.cgi?id=551724 & https://bugs.gentoo.org/show_bug.cgi?id=551928) so we are in progress to revert running in foreground and use the daemon again.

**Comment by arvidjaar** [comment]
> According to http://www.freedesktop.org/software/systemd/man/daemon.html (9) a daemonized service should connect /dev/null to standard input, output, and error. Can we do the same with systemd-udevd when running daemonized?

BTW udev actually does not implement several recommended steps like double-forking or synchronization between child and (grand-)parent. The latter means race window where udev is still not ready to process events but follow-up services may already be stated. Native unit is using notify so does not have this issue.

**Comment by kaysievers** [comment]
udev opens the sockets before forking, it will not return before that. Any events will be queued up in the socket, it does not need more synchronization.

**Comment by teg** [comment]
As @kaysievers said, there should be no problem with the current synchronization. However, to avoid having to deal with any obscure bugs like this again, I took a stab at going all the way and implemented the double-forking mode and the unifying the synchronization point in PR #229.

Could you verify that that works for you?

**Comment by Whissi** [comment]
I tried your PR. Looks good in general:
- I was able to boot my system using udevd from your PR, my interfaces for example were renamed like expected
- No open fd - restarting udev using _salt_ worked

But it is triggering an assertion on stop:

```
Assertion 'close_nointr(fd) != -EBADF' failed at /var/tmp/portage/sys-fs/udev-9999/work/systemd-9999/src/basic/util.c:309, function safe_close(). Aborting.
```

**Comment by teg** [comment]
Thanks for testing @Whissi. Could you have another go at it with PR #261? This is a much more minimal fix.

**Comment by teg** [comment]
Thanks for the comment @poettering, please test with #262 instead.

**Comment by Whissi** [comment]
Tested #262 - works. Thanks!

**Comment by dvdhrm** [comment]
Fixed by #262.
